added y argument to fit methods

vzaretsk opened the pull request:
Added optional y argument to fit methods of TransformerPipeline and DataFrameMapper. I'm using try/except TypeError to handle fit methods of pipeline transformers that don't accept a y argument. @dukebody

dukebody commented:
I like this! :)  It's simple enough, doesn't break existing code (I believe) and adds interesting functionality.

As you mentioned in #58, we should add some more tests that exercise this new functionality, as well as updating README documentation. Can you do this? If you need any pointers I'd be happy to help. :)

Thanks!

vzaretsk commented:
Yes, I'll do this. Let me work on it during the week and see how far I get. Do you have a Python 2 environment you can test the code in when it's ready? I don't have one and it would be nice if I didn't need to set one up. Thanks.

dukebody commented:
Sure, we use tox to test everything on Python 2 and 3.

2016-07-10 22:46 GMT+02:00 Vitaley Zaretskey <notifications@github.com>:

> Yes, I'll do this. Let me work on it during the week and see how far I
> get. Do you have a Python 2 environment you can test the code in when it's
> ready? I don't have one and it would be nice if I didn't need to set one
> up. Thanks.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/paulgb/sklearn-pandas/pull/59#issuecomment-231609951>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AACj4Z6yp3LvDuuhJbe6lGpXiVjVf-BSks5qUVoggaJpZM4JIu0Y>
> .
>

vzaretsk commented:
I added unit tests for the new functionality and updated the README. Everything is passing on Python 3. Let me know if you see any issues or anything else that needs to be added. Thanks.
@dukebody

dukebody commented on README.rst:
```diff
@@ -167,19 +167,36 @@ passing it as the ``default`` argument to the mapper:
-           [ 0.,  1.,  0.,  3.,  1.],
-           [ 1.,  0.,  0.,  5., -0.],
-           [ 0.,  0.,  1.,  4., -1.]])
+    >>> np.round(mapper4.fit_transform(data.copy()), 1)
```
Why does one need to pass "1" as second argument to this transform, and why is the output different from the previous case in the last column?

vzaretsk replied:
The "1" is the second argument of np.round. This test case was failing even though I don't think I modified anything that affected it. The issue seems to be that on my machine np.round(-0.3) equals "0.", not "-0." Changing it to round to 1 decimal place fixed the test case.

dukebody commented on sklearn_pandas/pipeline.py:
```diff
@@ -3,12 +3,37 @@
+    """
+    try:
+        return fit_method(X, y, **kwargs)
+    except TypeError:
```
What about:
```

except TypeError as e:
    if e.args and 'takes exactly' in e.args[0]:
        # fit takes only one argument
        return fit_method(X, **kwargs)
    # TypeError caused by some other unkown reason
    raise
```

I know it looks a bit hacky but I guess it will solve your warning above.

vzaretsk replied:
This is pretty clever, I hadn't thought of doing something like this before. Unfortunately when I tested it, the error message accompanying a TypeError varies between Python 2 and 3. On 2, it's "test_func() takes exactly 2 arguments (1 given)" but on 3 it's "test_func() missing 1 required positional argument:". It's probably safer to leave as is.

dukebody commented on tests/test_pipeline.py:
```diff
@@ -1,5 +1,13 @@
 import pytest
-from sklearn_pandas.pipeline import TransformerPipeline
+from sklearn_pandas.pipeline import TransformerPipeline, _call_fit
+from functools import partial
```
Imported but unused; please remove this import.

dukebody commented:
Good work! Tests pass both on Python 2.7 and 3.5.

Merging once you've addressed the inline comments I did on the existing commits.

vzaretsk commented:
Thanks. I answered the comments and made a new commit. @dukebody

dukebody commented:
All tests passing, so merged!
